shubhrajps/ecommerce-discounts
# [MAJOR] Ecommerce feature

### Description
This pull request introduces the core functionality for adding items to the cart and checking out orders in the ecommerce store. It also includes the implementation of discount codes for every nth order, which provides a 10% discount on the total order amount.

### Changes Made
- Added cart and checkout APIs to handle adding items to the cart and processing orders
- Implemented the logic to generate discount codes for every nth order
- Added validation in the checkout API to check if a discount code is valid before applying the discount
- Created admin APIs to:
- Generate discount codes based on the specified condition
- Retrieve statistics on purchased items, total purchase amount, discount codes, and total discounts
- Added tests to ensure the correctness of the implemented functionality

### Screenshots
**Homescreen**
<img width="1498" alt="Screenshot 2024-09-01 at 11 07 11 PM" src="https://github.com/user-attachments/assets/ecb3018b-0e0f-49c9-b83a-c5f1ac2abcff">

**Checkout**
<img width="1502" alt="Screenshot 2024-09-01 at 11 07 51 PM" src="https://github.com/user-attachments/assets/2a4db329-b2c2-417f-aca3-a074a315252d">

**Checkout**
<img width="1500" alt="Screenshot 2024-09-01 at 11 08 03 PM" src="https://github.com/user-attachments/assets/658f78fe-48cb-49bc-97b2-c23ec8fbc30a">
<img width="1498" alt="Screenshot 2024-09-01 at 11 08 30 PM" src="https://github.com/user-attachments/assets/c70c0adb-2927-441c-8aca-32237ede054d">


**Orders**
<img width="1364" alt="Screenshot 2024-09-01 at 11 08 48 PM" src="https://github.com/user-attachments/assets/2f851189-3704-4309-a19d-a81bdeaba882">


### Purpose
The purpose of this change is to provide the essential functionality for customers to add items to their cart and successfully place orders, with the added benefit of discount codes for every nth order. It also includes administrative APIs for managing discount codes and retrieving store statistics.

### How to Test
- Use the provided APIs to add items to the cart and place orders
- Verify that discount codes are generated correctly for every nth order
- Test the checkout API with valid and invalid discount codes
- Use the admin APIs to generate discount codes and retrieve store statistics

### Additional Notes
- The UI implementation is a stretch goal and not included in this PR
- Postman or REST client collections are provided for testing the APIs
